### ISSUE TITLE
Central certificate management instance

### ISSUE THREAD

**Issue by jloh** [caddyserver/caddy#4653]
This is a feature request.

I’m looking at rolling out a fleet of Caddy servers using either Consul or Redis to handle shared certificate management.

It would be cool if it was possible to only expose DNS credentials (needed for wildcard certificates) to a specific Caddy instance and leave certificate management up to that single instance whilst still letting the others in the “fleet” use and access the certificate. My use case is I want to limit what servers have access to my DNS credentials.

Basically; have one server renew and manage, the others just use to the serve traffic.

This is split from this forum topic https://caddy.community/t/central-certificate-management/15423/2

**Comment by mholt** [comment]
Thanks for opening an issue about this -- I think this is worth some discussion. I'm not 100% sure if there's anything useful here, but let's explore.

Question for you... what if there was a DNS provider module that "wrapped" any other DNS provider, and would do something to fetch the credentials and make them available as placeholders?

Unless, is your requirement that you can't have the credentials on the machine at all (even in memory)?

So when you say this, I want to be sure I'm on the same page about exactly what your requirements are:

> limit what servers have access

What _exactly_ do you mean? By "limit" how much are we talking, completely separate or behind authentication or... what is the threat model? By "server" do you mean the machine & OS itself? Or Caddy process specifically. And by "have access," in what way do you mean? Again, what is the threat model?

I'm willing to write up some feature or mode that can deactivate cert management on an instance but I need to be sure I totally understand the motivation and target the specific, real-life threat mode. (Ideally not a hypothetical one.)

**Comment by jloh** [comment]
Okay, great questions. I'll do my best to answer and try and outline my thinking but please let me know if I'm not making sense. I also want to say upfront that I completely understand the impact losing a private key for a certificate would have and am not trying to understate that. This is all about trying to limit exposure.

> limit what servers have access

I think for my use case limiting access is primarily about limiting which servers have to have the keys written to _disk_ in their config. So "server" in this sense means the OS. I think if Caddy were to reach out somewhere else (or get it from another host in the cluster?) that would be _okay_ since there is a bit of a leap of getting a key from memory vs reading it from a config file.

The threat model I have in my head is we have two sets of users that have access to the servers running Caddy. Our "ops" group and our standard users group. Ops group have access to all servers, standard users only have access to a smaller subset.

Currently I'm forced to expose DNS keys to all servers in a cluster running Caddy where I want to use a set of certificates, effectively exposing the credentials to all users that can access them.

If I was able to only expose the key to a specific servers config and the rest can learn it via another method (or ideally not need it at all!) then I can limit what users have easy access to it by having a "control" server that only Ops have access to for example.

I understand all users could technically get the private keys in this scenario and decrypt/MITM the traffic that way but to me there is a big difference between getting a private key and MITM'ing traffic and being able to get DNS API access and potentially redirect DNS _and_ MITM traffic.

Again, I understand the answer could be "you need other security protections in place, limit users etc" and I acknowledge that. I'm just trying to see if there is a way I can use Caddy in this scenario and limit scope.

---

Another thing to consider is most (at least in my experience) providers let you lock down API keys to specific IP addresses. If I could designate a specific Caddy instance to handle renewals for me and only provide _it_ with the API keys I can lock that key down to that machines specific IPs. This would again greatly limit exposure if the API key was lost.

This wouldn't fit into the "Caddy goes and gets config from elsewhere" model I don't think, but just something to consider I guess?

To me who can access DNS is critical since it has immediate public impact and with it a malicious user can really get any certificates they want, they don't even need to try and pry them from Caddy. So to summarise I guess my thread model is losing API keys to DNS and I'm just trying to limit where I expose that risk.

Hopefully this makes a _little_ bit of sense 😅

**Comment by mholt** [comment]
@jloh Thanks for the detailed answer, I find it very helpful/insightful.

While something like this could be doable ("use managed certs from storage, but don't _manage_ certs from storage"), what about something like this: https://caddy.community/t/new-acme-dns-libdns-plugin/15474/8?u=matt

ACME-DNS: https://github.com/joohoi/acme-dns

This project is essentially what you're asking for, and as of yesterday, there's a Caddy module that uses it: https://github.com/caddy-dns/acme-dns

This would be preferable if it suits your needs because then I don't have to change anything in Caddy or CertMagic :sweat_smile:

**Comment by gc-ss** [comment]
Matt, Thank You for pointing this new module out!

> I’m looking at rolling out a fleet of Caddy servers using either Consul or Redis to handle shared certificate management.
> My use case is I want to limit what servers have access to my DNS credentials

Very nice. You can use the service mesh ability of Consul to update the config of Caddy as well BTW, reducing the moving pieces. Depending on the Consul version and license you have, you might be able to use namespaces and ACLs to further tighten down. Let me know the details if you plan to go this route

Also, if you can use etcd instead (in which case you won't have the service mesh ability of Consul) - let me know and I will see what it takes to cook up a POC. In this scenario, etcd is a HA KV data store.

> bit of a leap of getting a key from memory vs reading it from a config file

Fair - we will use the Caddy admin API

> Currently I'm forced to expose DNS keys to all servers in a cluster running Caddy where I want to use a set of certificates, effectively exposing the credentials to all users that can access them.

Exactly. This was too much for us so we (hopefully?) solved it: https://caddy.community/t/new-acme-dns-libdns-plugin/15474/8?u=gcss

We are hoping to hear back and get feedback on whether it's a good solution :D

**Comment by mholt** [comment]
Since there has been no further feedback from @jloh and @gc-ss seems to have contributed a solution with the acme-dns plugin, I am going to close this issue for now, but feel free to continue the discussion. As far as I can tell, this is resolved; but if I am wrong just let me know.
